test/e2e/user_configured_cabundle_test.go

```diff
@@ -0,0 +1,692 @@
+		Subject:               pkix.Name{CommonName: commonName},
+		SignatureAlgorithm:    x509.SHA256WithRSA,
+		NotBefore:             time.Now().Add(-24 * time.Hour),
+		NotAfter:              time.Now().Add(24 * time.Hour),
+		SerialNumber:          serialNumber,
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		MaxPathLen:            0,
+		MaxPathLenZero:        true,
+	}
+
+	cert, err := createCertificate(root, root, &key.PublicKey, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return cert, key, nil
+}
+
+// generateServerCertificate generates and returns a server certificate and key
+// where the certificate is signed by the provided CA certificate.
+func generateServerCertificate(caCert *x509.Certificate, caKey *rsa.PrivateKey, cn string, dnsDomains []string) (*x509.Certificate, *rsa.PrivateKey, error) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	serialNumber, err := generateCertificateSerialNumber()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	template := &x509.Certificate{
+		Subject: pkix.Name{
+			CommonName:   cn,
+			Organization: []string{"OpenShift"},
+		},
+		SignatureAlgorithm:    x509.SHA256WithRSA,
+		NotBefore:             time.Now().Add(-24 * time.Hour),
+		NotAfter:              time.Now().Add(24 * time.Hour),
+		SerialNumber:          serialNumber,
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		DNSNames:              dnsDomains,
+	}
+
+	cert, err := createCertificate(template, caCert, &key.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return cert, key, nil
+}
+
+// buildTLSSecret returns ConfigMap object with the provided data.
+func buildConfigMap(name types.NamespacedName, data map[string]string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+		Data: data,
+	}
+}
+
+// buildTLSSecret returns a TLS secret object with the provided certificate and key.
+func buildTLSSecret(name types.NamespacedName, cert, key []byte) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+		Data: map[string][]byte{
+			corev1.TLSCertKey:       cert,
+			corev1.TLSPrivateKeyKey: key,
+		},
+		Type: corev1.SecretTypeTLS,
+	}
+}
+
+// buildServerDeployment returns deployment object with the provided TLS secret mounted as volume.
+func buildServerDeployment(name types.NamespacedName, labels map[string]string, port int32, secretName, caConfigMapName string) *appsv1.Deployment {
+	const (
+		containerName = "hello-world-server"
+		imageName     = "quay.io/bharath-b-rh/hello-world:latest"
```

Comment:
I think we need to try to use a standard image, and not one in a personal repo that could get deleted, or changed.

Some examples of images our existing tests use:

- `image-registry.openshift-image-registry.svc:5000/openshift/tools:latest`
- The router deployment's: `deployment.Spec.Template.Spec.Containers[0].Image`
- `quay.io/centos7/httpd-24-centos7` (this one I'm a little hesitant, but at least it's in a public repo)

Reply:
I can work on making source code public, and add more description to image in repo.

For reencrypt, since we need to verify service certificates, we need an https server and I was looking for alternatives. One I came across was apache httpd, but this requires enabling certain configurations for SSL handling, which would require building own image with changes or handling it through entrypoint script. 
Can we use `https://hub.docker.com/r/mendhak/http-https-echo/tags`?
Please suggest.

Reply:
I agree we should use `image-registry.openshift-image-registry.svc:5000/openshift/tools:latest` or `quay.io/centos7/httpd-24-centos7` if we can.  Alternatively, if the server code isn't too complex, we can build it into the ingress-operator executable, the way we did the the test servers in https://github.com/openshift/cluster-ingress-operator/tree/master/test.

Reply:
In any case, we cannot use an image from Docker Hub as we could be affected by rate limiting or other issues if we did.

Reply:
Is this something we can use `quay.io/centos7/httpd-24-centos7` and add the httpd configuration we need via a configmap?

Reply:
Removed the public image, instead now making use of the `ingress-operator` image fetched from the deployment resource to use the [test http server](https://github.com/openshift/cluster-ingress-operator/tree/master/test/http2) made available.